include/ginkgo/core/matrix/row_gatherer.hpp

```diff
@@ -0,0 +1,170 @@
+ * calls the respective LinOp's rowgatherer operation if the respective LinOp
+ * implements the RowGatherable interface. As such it only stores an array of
+ * the rowgatherer indices.
+ *
+ * @ingroup rowgatherer
+ * @ingroup matrix
+ * @ingroup LinOp
+ */
+template <typename IndexType = int32>
+class RowGatherer : public EnableLinOp<RowGatherer<IndexType>>,
+                    public EnableCreateMethod<RowGatherer<IndexType>> {
+    friend class EnableCreateMethod<RowGatherer>;
+    friend class EnablePolymorphicObject<RowGatherer, LinOp>;
+
+public:
+    using index_type = IndexType;
+
+    /**
+     * Returns a pointer to the array of row_gatherer_indices.
+     *
+     * @return the pointer to the row_gatherer_indices array.
+     */
+    index_type* get_row_gatherer_indices() noexcept
+    {
+        return row_gatherer_indices_.get_data();
+    }
+
+    /**
+     * @copydoc get_row_gatherer_indices()
+     *
+     * @note This is the constant version of the function, which can be
+     *       significantly more memory efficient than the non-constant version,
+     *       so always prefer this version.
+     */
+    const index_type* get_const_row_gatherer_indices() const noexcept
+    {
+        return row_gatherer_indices_.get_const_data();
+    }
+
+protected:
+    /**
+     * Creates an uninitialized RowGatherer arrays on the specified executor.
+     *
+     * @param exec  Executor associated to the LinOp
+     */
+    RowGatherer(std::shared_ptr<const Executor> exec)
+        : RowGatherer(std::move(exec), dim<2>{})
+    {}
+
+    /**
+     * Creates uninitialized RowGatherer arrays of the specified size.
+     *
+     * @param exec  Executor associated to the matrix
+     * @param size  size of the RowGatherable matrix
+     */
+    RowGatherer(std::shared_ptr<const Executor> exec, const dim<2>& size)
+        : EnableLinOp<RowGatherer>(exec, size),
+          row_gatherer_indices_(exec, size[0])
+    {}
+
+    /**
+     * Creates a RowGatherer matrix from an already allocated (and initialized)
+     * row gathering array
+     *
+     * @tparam IndicesArray  type of array of indices
+     *
+     * @param exec  Executor associated to the matrix
+     * @param size  size of the rowgatherer array.
+     * @param row_gatherer_indices array of rowgatherer array
+     *
+     * @note If `row_gatherer_indices` is not an rvalue, not an array of
+     * IndexType, or is on the wrong executor, an internal copy will be created,
+     * and the original array data will not be used in the matrix.
+     */
+    template <typename IndicesArray>
+    RowGatherer(std::shared_ptr<const Executor> exec, const dim<2>& size,
+                IndicesArray&& row_gatherer_indices)
+        : EnableLinOp<RowGatherer>(exec, size),
+          row_gatherer_indices_{
+              exec, std::forward<IndicesArray>(row_gatherer_indices)}
+    {
+        GKO_ASSERT_EQ(size[0], row_gatherer_indices_.get_num_elems());
+    }
```

Comment:
I know we do it like this in other places, but this makes the interface unnecessarily generic and forces it to be in the header.
```suggestion
    RowGatherer(std::shared_ptr<const Executor> exec, const dim<2>& size,
                Array<IndexType> row_gatherer_indices)
        : EnableLinOp<RowGatherer>(exec, size),
          row_gatherer_indices_{
              exec, std::move(row_gatherer_indices)}
    {
        GKO_ASSERT_EQ(size[0], row_gatherer_indices_.get_num_elems());
    }
```

Reply:
Ah nevermind, using this interface allows us to avoid a single copy if the array is on another executor than the RowGatherer